tuta-sdk/rust/src/crypto/entity_facade.rs

```diff
@@ -0,0 +1,304 @@
+            match element_value {
+                Ok((decrypted, ivs, errors)) => {
+                    mapped_decrypted.insert(key.to_string(), decrypted);
+                    mapped_decrypted.get("errors").unwrap().assert_dict().insert(key.to_string(), ElementValue::Dict(errors));
+                    mapped_decrypted.get("final_ivs").unwrap().assert_dict().insert(key.to_string(), ElementValue::Dict(ivs))
+                }
+                Err(err) => return Err(err)
+            };
+        }
+
+        let model_name = &type_model.name;
+        for (association_name, association_model) in type_model.associations.clone().into_iter() {
+            let association = self.map_associations(type_model, entity.to_owned(), session_key, model_name, &association_name, association_model);
+            match association {
+                Ok((mapped_association, errors)) => {
+                    mapped_decrypted.insert(association_name.to_string(), mapped_association);
+                    mapped_decrypted.get("errors").unwrap().assert_dict().insert(association_name.to_string(), ElementValue::Dict(errors));
+                }
+                Err(err) => return Err(err)
+            }
+        }
+
+        Ok(mapped_decrypted)
+    }
+
+    fn map_associations(&self, type_model: &TypeModel, entity: ParsedEntity, session_key: &AesKey, model_name: &String, association_name: &String, association_model: ModelAssociation) -> Result<(ElementValue, HashMap<String, ElementValue>), ApiCallError> {
+        let mut errors: HashMap<String, ElementValue> = Default::default();
+        let dependency = match association_model.dependency {
+            Some(dep) => dep,
+            None => type_model.app.clone()
+        };
+
+        let aggregate_type_model = match self.type_model_provider.get_type_model(&dependency, association_model.ref_type.as_str()) {
+            Some(type_model) => type_model,
+            // Undefined type model or type ref should be treated as panic as the system isn't
+            // capable of dealing with unknown types
+            None => return panic!("Undefined type_model {}", association_model.ref_type)
+        };
+
+        return if let AssociationType::Aggregation = association_model.association_type {
+            let association_data = entity.get(association_name);
+            if association_model.cardinality == Cardinality::ZeroOrOne && association_data.is_none() {
+                Ok((ElementValue::Null, errors))
+            } else if association_data.is_none() {
+                Err(ApiCallError::InternalSdkError { error_message: format!("Undefined aggregation {model_name}:{association_name}") })
+            } else if association_model.cardinality == Cardinality::Any {
+                let mut aggregate_vec: Vec<ElementValue> = Vec::new();
+                for aggregate in association_data.unwrap().assert_array().into_iter() {
+                    let aggregate_dict = aggregate.assert_dict();
```

Comment:
we shouldn't `assert_dict`, we should check and return error.

That's why I suggested matching on `(association_data, association_model.cardinality)`